sourced/ml/transformers/content2ids.py

```diff
@@ -0,0 +1,146 @@
+                    token=x[0],
+                    token_split=" ".join(TokenParser(min_split_length=1).split(x[0])),
+                    num_repos=x[1][0],
+                    num_files=x[1][1],
+                    num_occ=x[1][2]))
+        else:
+            return processed \
+                .map(lambda x: x[0]) \
+                .distinct() \
+                .map(lambda x: Row(
+                    token=x,
+                    token_split=" ".join(TokenParser(min_split_length=1).split(x))))
+
+    def reduce_rows(self, rows):
+        return rows \
+            .map(lambda x: (x[0], 1)) \
+            .reduceByKey(operator.add)
+
+    def _process_row(self, row):
+        self.names = []
+        repo_id = getattr(row, self.documents_column[0])
+        file_id = getattr(row, self.documents_column[1])
+        path = os.path.join(repo_id, file_id)
+        code = row.content
+        try:
+            lexer = get_lexer_by_name(self.linguist2pygments[row.lang][0])
+            pygments.highlight(code, lexer, self.FormatterProxy(callback=self.process_tokens))
+        except ClassNotFound:
+            lexer = get_lexer_by_name(self.linguist2pygments[row.lang][1])
+            pygments.highlight(code, lexer, self.FormatterProxy(callback=self.process_tokens))
+        except KeyError:
+            pass
+        for token in self.names:
+            yield token, (repo_id, path)
+
+    def process_tokens(self, tokens):
+        """
+        Filter tokens of type "Name" and which are splittable
+        according to :class: 'TokenParser' rules
+        """
+        for _type, token in tokens:
+            if _type[0] == "Name":
+                if self.split:
+                    if len(list(TokenParser(min_split_length=1).split(token))) > 1:
+                        self.names.append(token)
+                else:
+                    self.names.append(token)
+
+    def build_mapping(self):
+        """
+        Builds the mapping between linguist languages and pygments names for lexers.
+        """
+        with open("doc/languages.yml") as f:
+            all_languages = yaml.load(f)
+
+        linguist_langs = {}
+        for lang, specs in all_languages.items():
+            if specs["type"] == "programming":
+                linguist_langs[lang] = (set(specs.get("aliases", []) + [lang.lower()]), specs)
+
+        pygments_langs = set()
+        for lexer in pygments.lexers.LEXERS.values():
+            lang_declensions = [lang.lower() for lang in lexer[2]] + [lexer[1].lower()]
+            pygments_langs |= set(lang_declensions)
+
+        for lang in linguist_langs:
+            lang_names = linguist_langs.get(lang, (set(),))[0]
+            inter = list(lang_names.intersection(pygments_langs))
+            if inter:
+                self.linguist2pygments[lang] = inter
+
+    def save(self, id_rdd):
+        with gzip.open(self.output, "w") as g:
+            columns_names = ["token", "token_split"]
```

Comment:
I think it is a nice opportunity to use https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_csv.html and erase this tricky ad-hoc code